javascript/ql/src/semmle/javascript/CharacterEscapes.qll

```diff
@@ -0,0 +1,88 @@
+/**
+ * Provides classes for reasoning about character escapes in literals.
+ */
+
+import javascript
+
+module CharacterEscapes {
+  /**
+   * Provides sets of characters (as strings) with specific string/regexp characteristics.
+   */
+  private module Sets {
+    string ordinaryEscapeChars() { result = "fnrtvux0" }
+
+    string regexpCharRangeChars() { result = "bBcdDsSwW" }
+
+    string regexpBackreferenceChars() { result = "123456789k" }
+
+    string regexpMetaChars() { result = "^$*+?.()|{}[]-" }
+  }
+
+  /**
+   * Gets the `i'th` character of `raw`, which is preceeded by an uneven number of backslashes.
+   */
+  bindingset[raw]
+  string getAnEscapedCharacter(string raw, int i) {
+    result = raw.regexpFind("(?<=(^|[^\\\\])\\\\(\\\\{2}){0,10}).", _, i)
+  }
+
+  /**
+   * Gets a character in `n` that is preceeded by a single useless backslash.
+   *
+   * The character is the `i'th` character of `rawStringNode`'s raw string value.
+   */
+  string getAnIdentityEscapedCharacter(DataFlow::Node n, ASTNode rawStringNode, int i) {
+    exists(string raw, string escapableChars |
+      exists(string quote |
+        rawStringNode = n.asExpr() and
+        raw = rawStringNode.(StringLiteral).getRawValue() and
+        (
+          raw.charAt(0) = "'" and quote = "'"
+          or
+          raw.charAt(0) = "\"" and quote = "\""
+        )
+        or
+        rawStringNode = n.asExpr().(TemplateLiteral).getAnElement() and
+        raw = rawStringNode.(TemplateElement).getRawValue() and
+        quote = "`"
+      |
+        escapableChars = "b" + Sets::ordinaryEscapeChars() + quote + "\\\\"
```

Comment:
Could you explain the `+ "\\\\"` part? I understand that we don't want to treat `\\` as an identity escape, but `escapableChars` seems to be used as a set, so why add *two* backslashes?

Reply:
Good catch, I have moved `"\\"` into `Sets::ordinaryEscapeChars()`. An earlier iteration of this query used this in a regexp character class, so it had to be escaped twice in that context.